ProcessOptimizer/space/constraints.py

```diff
@@ -121,8 +129,158 @@ def rvs(self, n_samples=1, random_state=None):
+        while len(samples) < n_samples:
+            # Generate next step in the sequence
+            vec_comp = (seed + alpha*(j+1)) %1
+            j += 1
+            # Center these components on zero
+            vec_comp = vec_comp - 0.5
+            # Simulate lengths of each null_space vector to add to our point
+            vec_comp = vec_comp * sim_distance * 2
+            sample_candidate = point[:, None].T + ns @ vec_comp.T
+            # Generate the correct shape 
+            sample_candidate = sample_candidate[0]
+            # Check that the candidate is inside the original parameter space
+            inspace = [
+                (sample_candidate[i] >= self.space.bounds[i][0]) and
+                (sample_candidate[i] <= self.space.bounds[i][1])
+                for i in range(d)
+            ]
+            # Only accept the candidate if it is in our space
+            if all(inspace):
+                samples.append(sample_candidate)      
+        # Convert the list of arrays to a list of lists
+        samples = [arr.tolist() for arr in samples]
+        
+        # Create settings for the dimensions that are not part of the constraint
+        if d < self.space.n_dims:
+            remaining_dimensions = [
+                i for i in range(self.space.n_dims) 
+                if i not in self.sum_equals[0].dimensions
+            ]
+            # Convert our list of samples to an array
+            samples = np.array(samples)
+            # Expand the sample array to make space for settings of the 
+            # unconstrained dimensions
+            for i in remaining_dimensions:
+                samples = np.insert(
+                    samples,
+                    i, 
+                    np.zeros(len(samples)), 
+                    axis=1
+                )
+            # Convert back to list of lists
+            samples = samples.tolist()
+            
+            # Generate random settings across all factors
+            full_sample = np.array(
+                self.space.rvs(n_samples=n_samples, random_state=rng)
+            )
+            # Sort the settings in each column, which will ensure that the 
+            # unconstrained settings are distributed in a space-filling way too
+            full_sample.sort(axis=0)
+            full_sample = full_sample.tolist()
+            
+            # Overwrite the random setting values for the constrained factors
+            for j in remaining_dimensions:
+                for i in range(len(samples)):
+                    samples[i][j] = full_sample[i][j]
+            
+            # Shuffle the order of the samples, otherwise the unconstrained
+            # settings will be returned in a sorted order. Use seeding to 
+            # provide consistent initial samples
+            rng2 = np.random.default_rng(seed=42)
+            rng2.shuffle(samples)
+                    
+        return samples
+
+    def validate_sample(self, sample: list) -> bool:
```

Comment:
I am not sure Python 3.8 supports `list` as a type hint, so you might have to import `List` from `typing`. but perhaps only `list[class]` is a problem.
```suggestion
    def validate_sample(self, sample: List) -> bool:
```

Reply:
The correct usage is indeed with List instead of list.